src/lib/Live/Eval.hs

```diff
@@ -56,46 +68,183 @@ data CacheState = CacheState
+-- still the state as of the end of the previous message.
+evalSource :: SourceEvalState -> SourceContents -> DriverM ()
+evalSource freshEvalState source = do
+  -- Save the old state from the monad, because we need to kill or
+  -- reuse the worker from it.
+  oldTopState <- gets fst
+  modify $ onFst $ const freshEvalState
+  let UModule _ _ blocks = parseUModule Main source
+  (evaluated, remaining) <- tryEvalBlocksCached blocks
+  (reused, remaining') <- tryReuseWorker oldTopState remaining
+  remaining'' <- mapM makeNewBlockId remaining'
+  updateResultList $ map getNodeId $ evaluated ++ reused ++ remaining''
+  putPendingBlocks $ reused ++ remaining''
+  maybeLaunchWorker
+
+-- See which blocks already have completed values and reuse those.
+tryEvalBlocksCached :: [SourceBlock]
+                    -> DriverM ([WithId SourceBlock], [SourceBlock])
 tryEvalBlocksCached [] = return ([], [])
 tryEvalBlocksCached blocks@(block:rest) = do
-  (env, cache) <- get
+  ((SourceEvalState env _ _), cache) <- get
   case M.lookup (block, env) (evalCache cache) of
     Nothing -> return ([], blocks)
     Just (blockId, env') -> do
       let block' = WithId blockId block
-      updateTopState env'
+      putTopState env'
       (evaluated, remaining) <- tryEvalBlocksCached rest
       return (block':evaluated, remaining)
 
-evalBlock :: WithId SourceBlock -> DriverM ()
-evalBlock (WithId blockId block) = do
-  oldState <- gets fst
-  opts <- asks fst
-  (result, s) <- liftIO $ evalSourceBlockIO opts (withoutId oldState) block
+-- See whether the formerly active worker (if any) is still doing
+-- something useful given the list of blocks we are waiting to finish;
+-- if so reuse it, and if not kill it.
+tryReuseWorker :: SourceEvalState -> [SourceBlock]
+               -> DriverM ([WithId SourceBlock], [SourceBlock])
+tryReuseWorker (SourceEvalState _ w _) [] =
+  liftIO (forM_ w killThread) >> return ([], [])
+tryReuseWorker (SourceEvalState _ Nothing _) blocks =
+  return ([], blocks)
+tryReuseWorker (SourceEvalState oldEnv w (oldNext:_)) (next:rest) = do
+  curEnv <- gets $ envInState . fst
+  if (curEnv == oldEnv) && (withoutId oldNext == next) then do
+    -- Reuse the worker
+    putWorker w
+    return ([oldNext], rest)
+  else
+    liftIO (forM_ w killThread) >> return ([], next:rest)
+tryReuseWorker (SourceEvalState _ (Just _) []) _ =
+  error $ "Impossible: SourceEvalState contained a worker "
+  ++ "but no block for it to have been working on."
+
+-- If we have work to do but no worker doing it, launch such a worker.
+maybeLaunchWorker :: DriverM ()
+maybeLaunchWorker = do
+  (SourceEvalState curState curWorker curPending) <- gets fst
+  case curWorker of
+    (Just _) -> return ()
+    Nothing -> do
+      case curPending of
+        [] -> return ()
+        (next:_) -> do
+          opts <- asks fst
+          self <- lift ask
+          tid <- liftIO $ forkWorker opts curState next self
```

Comment:
Do we understand the cost of `liftIO`? We always have at most one worker, so if it's expensive we might want to make it persistent

Reply:
I assume you're concerned about the underlying `forkIO`?  AFAIK that's supposed to be cheap, because it creates a "lightweight thread" managed by the Haskell runtime (as opposed to `forkOS`, which actually forks a new OS-level thread).  The `liftIO` itself is inevitable, because we have to either fork a new worker or send a message to the persistent one, the latter of which is also an IO action.  It's also lifting `IO` into our `DriverM` stack, which shouldn't be bad?